flow/src/main/java/com/sequenceiq/flow/core/AbstractAction.java

```diff
@@ -74,24 +81,28 @@ public void execute(StateContext<S, E> context) {
             try {
                 Map<Object, Object> variables = context.getExtendedState().getVariables();
                 prepareExecution(payload, variables);
-                flowContext = createFlowContext(flowParameters, context, payload);
-                Object flowStartTime = variables.get(FLOW_START_TIME);
-                if (flowStartTime != null) {
-                    Object execTime = variables.get(FLOW_START_EXEC_TIME);
-                    long flowElapsed = (System.currentTimeMillis() - (long) flowStartTime) / MS_PER_SEC;
-                    long execElapsed = (System.currentTimeMillis() - (long) execTime) / MS_PER_SEC;
-                    String flowStateName = String.valueOf(variables.get(FLOW_STATE_NAME));
-                    long executionTime = Math.max(execElapsed, flowElapsed);
-                    LOGGER.debug("Stack: {}, flow state: {}, phase: {}, execution time {} sec", payload.getResourceId(),
-                            flowStateName, execElapsed > flowElapsed ? "doExec" : "service", executionTime);
-                    metricService.submit(FlowMetricType.FLOW_STEP, executionTime, Map.of("name", flowStateName.toLowerCase()));
+                String flowStateName = String.valueOf(variables.get(FLOW_STATE_NAME));
+                Span activeSpan = tracer.activeSpan();
+                String operationName = context.getEvent().name();
+                SpanContext spanContext = flowParameters.getSpanContext();
+                if (TracingUtil.isActiveSpanReusable(activeSpan, spanContext, operationName)) {
+                    flowContext = createFlowContext(flowParameters, context, payload);
+                    doExecute(context, payload, flowContext, variables, flowStateName);
+                } else {
+                    Span span = TracingUtil.getSpan(tracer, operationName, spanContext, flowParameters.getFlowId(),
+                            null, flowParameters.getFlowTriggerUserCrn());
+                    if (spanContext == null) {
+                        spanContext = span.context();
+                    }
```

Comment:
if we create above a span with spanContext, do we really need this if? we could go with
`flowParameters.setSpanContext(span.context());`
if I'm not mistaken

Reply:
Yes it is needed because we always reuse the original and first span context. This way the trace will be usable, otherwise spans would be recursively coming from each other which is untrue for the trace. 
If a context is present we use it, else if this is the first step of a flow, we have to use the current span's context.